pyomo/contrib/pyros/pyros_algorithm_methods.py

```diff
@@ -47,6 +47,150 @@ def update_grcs_solve_data(
+
+
+def evaluate_and_log_component_stats(model_data, separation_model, config):
+    """
+    Evaluate and log model component statistics.
+    """
+    IterationLogRecord.log_header_rule(config.progress_logger.info)
+    config.progress_logger.info("Model statistics:")
+    # print model statistics
+    dr_var_set = ComponentSet(
+        chain(
+            *tuple(
+                indexed_dr_var.values()
+                for indexed_dr_var in model_data.working_model.util.decision_rule_vars
+            )
+        )
+    )
+    first_stage_vars = [
+        var
+        for var in model_data.working_model.util.first_stage_variables
+        if var not in dr_var_set
+    ]
+
+    # account for epigraph constraint
+    sep_model_epigraph_con = getattr(separation_model, "epigraph_constr", None)
+    has_epigraph_con = sep_model_epigraph_con is not None
+
+    num_fsv = len(first_stage_vars)
+    num_ssv = len(model_data.working_model.util.second_stage_variables)
+    num_sv = len(model_data.working_model.util.state_vars)
+    num_dr_vars = len(dr_var_set)
+    num_vars = int(has_epigraph_con) + num_fsv + num_ssv + num_sv + num_dr_vars
+
+    num_uncertain_params = len(model_data.working_model.util.uncertain_params)
+
+    eq_cons = [
+        con
+        for con in model_data.working_model.component_data_objects(
+            Constraint, active=True
+        )
+        if con.equality
+    ]
+    dr_eq_set = ComponentSet(
+        chain(
+            *tuple(
+                indexed_dr_eq.values()
+                for indexed_dr_eq in model_data.working_model.util.decision_rule_eqns
+            )
+        )
+    )
+    num_eq_cons = len(eq_cons)
+    num_dr_cons = len(dr_eq_set)
+    num_coefficient_matching_cons = len(
+        getattr(model_data.working_model, "coefficient_matching_constraints", [])
+    )
+    num_other_eq_cons = num_eq_cons - num_dr_cons - num_coefficient_matching_cons
+
+    # get performance constraints as referenced in the separation
+    # model object
+    new_sep_con_map = separation_model.util.map_new_constraint_list_to_original_con
+    perf_con_set = ComponentSet(
+        new_sep_con_map.get(con, con)
+        for con in separation_model.util.performance_constraints
+    )
+    is_epigraph_con_first_stage = (
+        has_epigraph_con and sep_model_epigraph_con not in perf_con_set
+    )
+    working_model_perf_con_set = ComponentSet(
+        model_data.working_model.find_component(new_sep_con_map.get(con, con))
+        for con in separation_model.util.performance_constraints
+        if con is not None
+    )
+
+    num_perf_cons = len(separation_model.util.performance_constraints)
+    num_fsv_bounds = sum(
+        int(var.lower is not None) + int(var.upper is not None)
+        for var in first_stage_vars
+    )
+    ineq_con_set = [
+        con
+        for con in model_data.working_model.component_data_objects(
+            Constraint, active=True
+        )
+        if not con.equality
```

Comment:
FYI.  This is a longstanding issue in Pyomo, but `con.equality` is only True if the constraint was declared as an Equality constraint.  Inequalities with equal upper/lower bounds still return False.